### ISSUE TITLE
Docker web socket/login issues

### ISSUE THREAD

**Issue by Joshuah143** [TabbycatDebate/tabbycat#1932]
After completing the setup of the develop branch on docker using the `docker compose up` command. I experienced an issue with login, it would take my credentials but not authenticate them and put me back to the page accessed immediately before the login request in a logged-out state. 

The logs are as follows:
> tabby-web-1 | 22:13:41 asgi.1 | 2021-12-28 09:13:41,348 - daphne.ws_protocol - DEBUG - WebSocket closed for ['172.21.0.1', 0]
> tabby-web-1 | 22:13:48 asgi.1 | 2021-12-28 09:13:48,965 - daphne.http_protocol - DEBUG - Upgraded connection ['172.21.0.1', 0] to WebSocket
> tabby-web-1 | 22:13:48 asgi.1 | 2021-12-28 09:13:48,971 - daphne.server - INFO - failing WebSocket opening handshake ('Access denied')
> tabby-web-1 | 22:13:48 asgi.1 | 2021-12-28 09:13:48,972 - daphne.server - WARNING - dropping connection to peer unix:None with abort=False: Access denied
> tabby-web-1 | 22:13:48 asgi.1 | 2021-12-28 09:13:48,972 - daphne.ws_protocol - DEBUG - WebSocket ['172.21.0.1', 0] rejected by application
> tabby-web-1 | 22:13:48 nginx.1 | 172.21.0.1 - - [27/Dec/2021:22:13:48 +0000] "GET /ws/bp88team/action_logs/ HTTP/1.1" 403 5 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.1 Safari/605.1.15"

Notably, this is running with HTTP, not HTTPS.

Versions:
Docker: 1.4.12
Host: MacOS 12.0.1
Tabbycat: 2.7.0

**Comment by czlee** [comment]
Just confirming that I can reproduce this on MacOS 11.6.1, Docker 20.10.10 b485636. My logs just say (abridged)
```
[01/Jan/2022:03:13:00 +0000] "GET /accounts/login/ HTTP/1.1" 200 3971 "http://localhost:8000/"
[01/Jan/2022:03:13:05 +0000] "POST /accounts/login/ HTTP/1.1" 302 0 "http://localhost:8000/accounts/login/"
[01/Jan/2022:03:13:06 +0000] "GET / HTTP/1.1" 200 4952 "http://localhost:8000/accounts/login/"
```
Seems to happen most of the time, but not all of the time.

I also tried it on Windows 11 (21H2, 21000.376), Docker 20.10.10 dea9396, on WSL 2 (Ubuntu 20.04.3), and can't see the issue there.

The `django_session` table is empty on the Mac when it's not working, and has one entry on WSL when it is working. So it appears not to be creating a session at all. (To check this: `docker compose exec db psql -U tabbycat` to enter the Postgres shell, then `SELECT * FROM django_session;`.)

**Comment by czlee** [comment]
Occasionally the login on macOS appears to succeed. But even then, `django_session` remains empty (or at least becomes empty before I get to look at it), and I'm reprompted to log in when I go to any next page.

I got the postgres container to log all queries, and on macOS it doesn't appear to run the `INSERT INTO django_session` query when it logs in. On WSL it does. So for whatever reason on macOS it's not even inserting anything into the sessions table…

**Comment by czlee** [comment]
Okay, I think I've got it. Redis (our cache) isn't currently enabled by default in Docker. Since we're using [cached sessions](https://docs.djangoproject.com/en/4.0/topics/http/sessions/#using-cached-sessions), this meant sessions weren't being saved at all. I'm not 100% on this explanation though, @philipbelesky might have more insight? Anyway, I've put a fix in #1951, on the `bug/docker-redis` branch. @Joshuah143, do you want to see if that resolves the issue for you?

To explain my observations above: For some reason (I'm not sure why) my WSL environment was also taking in my local.py settings overrides, which was setting `SESSION_ENGINE` to `"django.contrib.sessions.backends.cached_db"`. Then the cache wouldn't be working, but it would fall back to the database, which was working. When I removed my local.py file, I could reproduce the issue on WSL. So the OS was a red herring. And it's not actually meant to be logging sessions into the sessions table in the database, so that was a red herring too.

**Comment by czlee** [comment]
@Joshuah143 just a heads-up that I've bumped the Python version in Docker up to 3.8, it shouldn't really matter but I think it won't take effect until you rebuild the Docker container (`docker compose build web worker`).
